### PR TITLE
feat(launch): Creates a `--build` flag for building then pushing the image to a queue

### PR DESCRIPTION
**Pull request by gtarpenning**
Fixes WB-10393

Description
-----------
When passing in the new `--build` flag, launch uses a URI to build a docker image and create a “Job” artifact, then pushes a run queue item that uses this job artifact to run the associated job.

Overwrites a given URI, requires the `--queue` flag to be set. 

Testing
-------
New test using the relay server: `tests/unit_tests/test_launch_add.py`

### CODE REVIEW

**Comment by gtarpenning** [comment]
Looks like my handing of the Run object is not sufficient. Perhaps we can chat about this @KyleGoyette on Tuesday?

File: wandb/sdk/launch/_project_spec.py [hunk 1]
```diff
@@ -39,6 +41,14 @@ class LaunchSource(enum.IntEnum):
     JOB: int = 5
 
 
+class EntrypointDefaults(List[str]):
+    PYTHON = ["python", "main.py"]
+
+
+class LaunchType(str):
```
Comment (hu-po) [inline wandb/sdk/launch/_project_spec.py:48]: This seems kinda intense, might be simpler and easier to just have a variable `JOB_LAUCH_TYPE` rather than this class that has a single property

File: wandb/sdk/launch/_project_spec.py [hunk 2]
```diff
@@ -501,3 +511,84 @@ def create_metadata_file(
...
+    # circular dependency, TODO: chat with @Kyle
+    from wandb.sdk.launch.builder.loader import load_builder
+
+    assert launch_project.uri, "To build an image on queue a URI must be set."
+
+    repository: Optional[str] = launch_config.get("url")
+    builder_config = {"type": build_type}
+
+    docker_args = {}
+    if launch_project.python_version:
```
Comment (hu-po) [inline wandb/sdk/launch/_project_spec.py:540]: `is not None` ? maybe some type checking could be useful too

File: wandb/sdk/launch/_project_spec.py [hunk 3]
```diff
@@ -501,3 +511,84 @@ def create_metadata_file(
             },
             f,
         )
+
+
+def build_image_from_project(
```
Comment (gtarpenning) [inline wandb/sdk/launch/_project_spec.py:516]: Move into `build.py` or a `build_utils.py` files

File: wandb/sdk/launch/_project_spec.py [hunk 4]
```diff
@@ -501,3 +511,84 @@ def create_metadata_file(
...
+    updates launch_project with the newly created docker image uri and
+    returns the uri
+    """
+    # circular dependency, TODO: chat with @Kyle
+    from wandb.sdk.launch.builder.loader import load_builder
+
+    assert launch_project.uri, "To build an image on queue a URI must be set."
+
+    repository: Optional[str] = launch_config.get("url")
+    builder_config = {"type": build_type}
```
Comment (gtarpenning) [inline wandb/sdk/launch/_project_spec.py:537]: TODO: Add support for reading from config to use other builder

File: wandb/sdk/launch/_project_spec.py [hunk 5]
```diff
@@ -501,3 +511,84 @@ def create_metadata_file(
...
+    output_types = TypeRegistry.type_of(run.summary._as_dict()).to_json()
+
+    import pkg_resources
+
+    installed_packages_list = sorted(
+        f"{d.key}=={d.version}" for d in iter(pkg_resources.working_set)
+    )
+
+    source_info = {
+        "_version": "v0",
```
Comment (gtarpenning) [inline wandb/sdk/launch/_project_spec.py:580]: pack into job creator class

Comment (gtarpenning) [reply wandb/sdk/launch/_project_spec.py:580]: DRY out with similar code in `wandb_run.py`

File: wandb/sdk/launch/_project_spec.py (hunk 1 above)
Comment (gtarpenning) [reply wandb/sdk/launch/_project_spec.py:49]: I guess this only makes sense if we think there will ever be a need/want to support other default entry points. Like if the container is a bash operator or something and we want to have a `['sh', 'main.sh']` or something. If not then yeah I totally agree this is verbose for no reason.

Comment (KyleGoyette) [reply wandb/sdk/launch/_project_spec.py:49]: Where this is being used I think this is unnecessary. We can just write "launch_job", if I understand right the created Job should be entrypoint agnostic right?

File: wandb/sdk/wandb_run.py [hunk 6]
```diff
@@ -2194,6 +2203,13 @@ def _create_image_job(
         artifact = self.log_artifact(job_artifact)
         return artifact
 
+    def log_job_artifact(self, docker_image_name: str) -> Artifact:
```
Comment (KyleGoyette) [inline wandb/sdk/wandb_run.py:2206]: since this function will only create an image job, lets put "image" in the function name. Also prefix this with an underscore so users don't try to call it themselves

File: wandb/sdk/launch/_project_spec.py [hunk 7]
```diff
@@ -501,3 +512,24 @@ def create_metadata_file(
             },
             f,
         )
+
+
+def log_job_from_run(run: Run, docker_image_uri: str) -> str:
```
Comment (KyleGoyette) [inline wandb/sdk/launch/_project_spec.py:517]: I believe this function is now unused?

File: tests/unit_tests/tests_launch/test_launch_add.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    """
+    queue = "test_queue"
+    proj = "test_launch_build"
+    args = [
+        "https://github.com/gtarpenning/wandb-launch-test",
+        f"--project={proj}",
+        f"--entity={user}",
+        f"--queue={queue}",
+        "--job=oops",
+        "--build",
```
Comment (KyleGoyette) [inline tests/unit_tests/tests_launch/test_launch_add.py:38]: mock `docker.build_image` to just return a dummy uri this should work

Comment (KyleGoyette) [reply tests/unit_tests/tests_launch/test_launch_add.py:38]: validate that job artifact exists on server

File: tests/unit_tests/tests_launch/test_launch_add.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import time
+
+import pytest
+import wandb
+from wandb.apis.public import Api as PublicApi
+from wandb.sdk.internal.internal_api import Api as InternalApi
+from wandb.sdk.launch.launch_add import launch_add
+
+
+@pytest.mark.timeout(300)  # builds a container
```
Comment (gtarpenning) [inline tests/unit_tests/tests_launch/test_launch_add.py:11]: lol this is false, TODO: remove

File: tests/unit_tests/tests_launch/test_launch_add.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
+import os
+import time
+
+import wandb
+from wandb.apis.public import Api as PublicApi
+from wandb.sdk.internal.internal_api import Api as InternalApi
+from wandb.sdk.launch.launch_add import launch_add
+
+
+def test_launch_build_push_job(relay_server, runner, user, monkeypatch):
```
Comment (gtarpenning) [inline tests/unit_tests/tests_launch/test_launch_add.py:10]: add `import pytest` 
delete `runner`

File: wandb/sdk/launch/builder/build.py [hunk 11]
```diff
@@ -499,3 +509,75 @@ def _create_docker_build_ctx(
 def join(split_command: List[str]) -> str:
     """Return a shell-escaped string from *split_command*."""
     return " ".join(shlex.quote(arg) for arg in split_command)
+
+
+def make_image_uri(
```
Comment (KyleGoyette) [inline wandb/sdk/launch/builder/build.py:514]: nit: since we're building the image here, we can name this function `build_image_with_builder` or something

File: wandb/sdk/launch/builder/build.py [hunk 12]
```diff
@@ -499,3 +509,75 @@ def _create_docker_build_ctx(
...
+        repository = launch_config.get("url")
+
+    if launch_project.python_version:
+        docker_args["python_version"] = launch_project.python_version
+
+    if launch_project.cuda_version:
+        docker_args["cuda_version"] = launch_project.cuda_version
+
+    if launch_project.docker_user_id:
+        docker_args["user"] = str(launch_project.docker_user_id)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/builder/build.py:563]: this too should not be a docker arg

File: wandb/sdk/launch/builder/build.py [hunk 13]
```diff
@@ -499,3 +509,75 @@ def _create_docker_build_ctx(
...
+
+    wandb.termlog(f"{LOG_PREFIX}Building docker image from uri source.")
+    launch_project = fetch_and_validate_project(launch_project, api)
+    builder = load_builder(builder_config)
+    entry_point: EntryPoint = launch_project.get_single_entry_point() or EntryPoint(
+        name="main.py",
```
Comment (KyleGoyette) [inline wandb/sdk/launch/builder/build.py:569]: This should be `EntrypointDefaults.PYTHON[-1]`

File: wandb/sdk/launch/launch_add.py [hunk 14]
```diff
@@ -152,6 +160,23 @@ def _launch_add(
...
     )
+
+    if build:
+        if launch_spec.get("job") is not None:
+            wandb.termwarn("Build doesn't support setting a job. Overwriting job.")
+            launch_spec["job"] = None
+
+        launch_project = create_project_from_spec(launch_spec, api)
+        docker_image_uri = build_image_from_project(
+            launch_project, api, build_type="docker"
```
Comment (KyleGoyette) [inline wandb/sdk/launch/launch_add.py:171]: Lets try to source the `build_type` from the `LAUNCH_CONFIG_FILE` and default to docker. This will make supporting other builders smoother in the future

File: wandb/sdk/wandb_run.py [hunk 15]
```diff
@@ -2046,29 +2046,38 @@ def _on_ready(self) -> None:
...
+            output_types,
+        ) = self._make_job_source_reqs()
 
         for job_creation_function in [
             self._create_repo_job,
             self._create_artifact_job,
             self._create_image_job,
         ]:
-            artifact = job_creation_function(
+            artifact = job_creation_function(  # type: ignore
```
Comment (KyleGoyette) [inline wandb/sdk/wandb_run.py:2080]: why is this needed?

File: wandb/sdk/launch/builder/build.py [hunk 16]
```diff
@@ -499,3 +509,75 @@ def _create_docker_build_ctx(
...
+    updates launch_project with the newly created docker image uri and
+    returns the uri
+    """
+    assert launch_project.uri, "To build an image on queue a URI must be set."
+
+    builder_config = {"type": build_type}
+    docker_args = {}
+    repository = None
+    if launch_config:
+        repository = launch_config.get("url")
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/builder/build.py:554]: the cli ARG should take precedence over any default, and otherwise we should leave repository as None so it isn't attempted to be pushed to a non existent repo

**Review by KyleGoyette — CHANGES_REQUESTED** [review]
Looking good, a few changes requested. Thanks for dealing with the tough testing system!

File: wandb/sdk/launch/builder/build.py (hunk 11 above)
Comment (gtarpenning) [reply wandb/sdk/launch/builder/build.py:514]: Yeah I made this function for more easily isolating this step while testing, totally did not think about the name for more than about 3 seconds.

File: wandb/sdk/wandb_run.py (hunk 15 above)
Comment (gtarpenning) [reply wandb/sdk/wandb_run.py:2080]: flake 8: `error: Cannot call function of unknown type  [operator]`

File: wandb/sdk/launch/builder/build.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply wandb/sdk/launch/builder/build.py]: The jobs image uri should be `repo + ":" + launch_project.image_tag`

File: wandb/sdk/launch/builder/build.py (hunk 13 above)
Comment (gtarpenning) [reply wandb/sdk/launch/builder/build.py:588]: I did not address this because I'm not sure I exactly know what you mean.

Comment (KyleGoyette) [reply wandb/sdk/launch/builder/build.py:588]: since `EntrypointDefaults` is a list = `["python", "main.py"]` we should just use the last value of the array instead of hard coding `main.py` so it changes with the default entrypoint.

File: wandb/sdk/wandb_run.py (hunk 15 above)
Comment (KyleGoyette) [reply wandb/sdk/wandb_run.py:2102]: Makes sense!

**Review by KyleGoyette — CHANGES_REQUESTED** [review]
I think there's some confusion about the repository variable. The repository in build points to the plce where the image will be stored. We should make it copnfigurable through a CLI arg as well.

File: wandb/sdk/launch/builder/build.py [hunk 17]
```diff
@@ -499,3 +510,93 @@ def _create_docker_build_ctx(
...
+    object, with an optional launch_config to set git-things like repository
+    which is used in naming the output docker image, and build_type defaulting
+    to docker (but could be used to build kube resource jobs w/ "kaniko")
+
+    updates launch_project with the newly created docker image uri and
+    returns the uri
+    """
+    assert launch_project.uri, "To build an image on queue a URI must be set."
+
+    repository = launch_project.git_repo
```
[the remark this answers is not in the document]
Comment (gtarpenning) [reply wandb/sdk/launch/builder/build.py:575]: I don't think you can actually set the repository from the CLI, only the git_repo, which was where I got confused I think.

File: wandb/cli/cli.py [hunk 18]
```diff
@@ -1101,6 +1101,12 @@ def _parse_settings(settings):
     help="Flag to build an image with CUDA enabled. If reproducing a previous wandb run that ran on GPU, a CUDA-enabled image will be "
     "built by default and you must set --cuda=False to build a CPU-only image.",
 )
+@click.option(
+    "--build",
+    "-b",
+    is_flag=True,
+    help="Allow users to build image on queue then pushes a Job artifact. requires --queue to be set, default is false.",
```
[the remark this answers is not in the document]
Comment (gtarpenning) [reply wandb/cli/cli.py]: Should I include that change in this PR? rn I left in a TODO where the build code would need to include the repository param from either the LaunchProject or I guess we could pass it in directly.

File: wandb/sdk/launch/builder/build.py [hunk 19]
```diff
@@ -499,3 +510,93 @@ def _create_docker_build_ctx(
...
+    updates launch_project with the newly created docker image uri and
+    returns the uri
+    """
+    assert launch_project.uri, "To build an image on queue a URI must be set."
+
+    # TODO(gst): launch_project.repository support w/ CLI arg (pref over config)
+    repository = None
+    if launch_config:
+        repository = launch_config.get("repository")
+    docker_args, builder_config, _ = construct_builder_args(launch_config)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/builder/build.py:579]: The third argument returned here is the registry config, which contains a key `url` this value is the repository value you want to use.
Lets use that if the repository is not set by the CLI arg or launch config.

**Review by KyleGoyette — CHANGES_REQUESTED** [review]
We need to get and use the repository as listed in the LAUNCH_CONFIG_FILE file. This is contained in the registry_config, which is the third value returned by `construct_builder_args`

File: wandb/sdk/wandb_run.py [hunk 20]
```diff
@@ -2222,6 +2234,17 @@ def _create_image_job(
         artifact = self.use_artifact(job_artifact)
         return artifact
 
+    def _log_job_artifact_with_image(self, docker_image_name: str) -> Artifact:
+        packages, in_types, out_types = self._make_job_source_reqs()
+        job_artifact = self._create_image_job(
```
Comment (KyleGoyette) [inline wandb/sdk/wandb_run.py:2239]: Add a call to `log_artifact` in here, and remove the calls to `use_artifact` in each of the `_create_*_job` functions and put a use_artifact call in `_log_job`

**Comment by gtarpenning** [comment]
@KyleGoyette  holy shit we did the thing  ✅

File: wandb/sdk/wandb_run.py [hunk 21]
```diff
@@ -2219,8 +2231,21 @@ def _create_image_job(
...
+
+        return job_artifact
+
+    def _log_job_artifact_with_image(self, docker_image_name: str) -> Artifact:
+        packages, in_types, out_types = self._make_job_source_reqs()
+        job_artifact = self._create_image_job(
+            in_types, out_types, packages, docker_image_name
+        )
+
+        artifact = self.log_artifact(job_artifact)
```
Comment (KyleGoyette) [inline wandb/sdk/wandb_run.py:2243]: This should be `use_artifact`

**Review by KyleGoyette — COMMENTED** [review]
One change, then I think it LGTM

File: wandb/sdk/wandb_run.py (hunk 21 above)
Comment (KyleGoyette) [reply wandb/sdk/wandb_run.py:2243]: Nevermind

Comment (KyleGoyette) [reply wandb/sdk/wandb_run.py:2243]: It should be log, since we want the first case of the call with use to be the actual run

**Review by KyleGoyette — APPROVED** [review]
LGTM
